### PR TITLE
[topology] track connection time of neighbors

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces a mechanism to track the connection time of
neighbors as the number of seconds since the last successful link
establishment with the neighbor. This feature requires  the
configuration `OPENTHREAD_CONFIG_UPTIME_ENABLE` to be enabled which
is now enabled by default on FTD builds. The connection time of each
neighbor is available in the `otNeighborInfo` and `otChildInfo`
structures as the `mConnectionTime` member variable. A new CLI
command, `neighbor conntime`, has been added to display the
connection time and age of all neighbors.

### CODE REVIEW

File: include/openthread/thread.h [hunk 1]
```diff
@@ -86,11 +86,14 @@ typedef struct otLinkModeConfig
...
  *
+ * `mUptime` requires `OPENTHREAD_CONFIG_UPTIME_ENABLE`.
+ *
  */
 typedef struct
 {
     otExtAddress mExtAddress;           ///< IEEE 802.15.4 Extended Address
-    uint32_t     mAge;                  ///< Time last heard
+    uint32_t     mAge;                  ///< Seconds since last heard
+    uint32_t     mUptime;               ///< Seconds since link establishment
```
Comment (jwhui) [inline include/openthread/thread.h:96]: Rather than `uptime`, maybe refer to it as `ConnectionTime`?

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -5364,6 +5364,39 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+     * Print the uptime and age of neighbors. Info per neighbor:
+     * - RLOC16
+     * - Extended MAC address
+     * - Age (seconds since last heard from neighbor)
+     * - Uptime (seconds since link establishment with neighbor)
+     */
+    else if (aArgs[0] == "uptime")
+    {
+        static const char *const kUptimeTableTitles[]       = {"RLOC16", "Extended MAC", "Age", "Uptime"};
+        static const uint8_t     kUptimeTableColumnWidths[] = {8, 18, 16, 16};
```
Comment (jwhui) [inline src/cli/cli.cpp:5388]: 16 chars seems larger than needed for a 32-bit value.

16 chars seems smaller than needed for a 64-bit value.

File: include/openthread/thread.h (hunk 1 above)
Comment (abtink) [reply include/openthread/thread.h:96]: Thanks. Will rename in next push.

File: src/cli/cli.cpp (hunk 2 above)
Comment (abtink) [reply src/cli/cli.cpp:5388]: Will change in next push.
I may actually change the table to use a nicer format (`"<hh>:<mm>::<ss>"`) for durations (to make it easier to read).

Comment (jwhui) [reply src/cli/cli.cpp:5388]: `<hh>:<mm>::<ss>` would be nicer!

Comment (abtink) [reply src/cli/cli.cpp]: Added this format in the new pushed commit.

File: src/core/common/uptime.hpp [hunk 3]
```diff
@@ -90,13 +90,38 @@ class Uptime : public InstanceLocator, private NonCopyable
      * This method converts an uptime value (number of milliseconds) to a human-readable string.
      *
      * The string follows the format "<hh>:<mm>:<ss>.<mmmm>" for hours, minutes, seconds and millisecond (if uptime is
-     * shorter than one day) or "<dd>d.<hh>:<mm>:<ss>.<mmmm>" (if longer than a day).
+     * shorter than one day) or "<dd>d.<hh>:<mm>:<ss>.<mmmm>" (if longer than a day). @p aIncludeMsec can be used
+     * to determine whether `.<mmm>` milliseconds to be included or omitted in the resulting string.
```
Comment (jwhui) [inline src/core/common/uptime.hpp:94]:
```suggestion
     * to determine whether `.<mmm>` milliseconds is included or omitted in the resulting string.
```

File: src/core/common/uptime.hpp [hunk 4]
```diff
@@ -90,13 +90,38 @@ class Uptime : public InstanceLocator, private NonCopyable
...
+     * @param[in,out] aWriter        A `StringWriter` to append the converted string to.
+     * @param[in]     aIncludeMsec   Whether to include `.<mmm>` milliseconds in the string.
+     *
+     */
+    static void UptimeToString(uint64_t aUptime, StringWriter &aWriter, bool aIncludeMsec);
+
+    /**
+     * This static method converts a given uptime as number of milliseconds to number of seconds.
+     *
+     * @param[in] aUptimeInMilliseconds    Uptime in milliseconds (as `uint64_t`)
```
Comment (jwhui) [inline src/core/common/uptime.hpp:106]:
```suggestion
     * @param[in] aUptimeInMilliseconds    Uptime in milliseconds (as `uint64_t`).
```

File: src/core/common/uptime.hpp [hunk 5]
```diff
@@ -90,13 +90,38 @@ class Uptime : public InstanceLocator, private NonCopyable
...
+     */
+    static uint32_t MsecToSec(uint64_t aUptimeInMilliseconds)
+    {
+        return static_cast<uint32_t>(aUptimeInMilliseconds / 1000u);
+    }
+
+    /**
+     * This static method converts a given uptime as number of seconds to number of milliseconds.
+     *
+     * @param[in] aUptimeInSeconds    Uptime in seconds (as `uint32_t`)
```
Comment (jwhui) [inline src/core/common/uptime.hpp:119]:
```suggestion
     * @param[in] aUptimeInSeconds    Uptime in seconds (as `uint32_t`).
```

File: src/cli/cli.cpp [hunk 6]
```diff
@@ -5352,6 +5352,82 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+         * @code
+         * neighbor conntime list
+         * 0x8401 1a28be396a14a318 age:63 conn-time:644
+         * 0x5c00 723ebf0d9eba3264 age:23 conn-time:852
+         * 0xe800 ce53628a1e3f5b3c age:23 conn-time:180
+         * @endcode
+         * @par
+         * Print contention time and age of neighbors.
+         * This command is similar to `neighbor conntime`, but it displays the information in a list format. The age
+         * and connection time are both displayed in seconds
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5388]: Missing period at the end.

File: src/cli/cli.cpp [hunk 7]
```diff
@@ -5352,6 +5352,82 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+#if OPENTHREAD_CONFIG_UPTIME_ENABLE
+    /**
+     * @cli neighbor conntime
+     * @code
+     * neighbor conntime
+     * | RLOC16 | Extended MAC     | Last Heard (Age) | Connection Time  |
+     * +--------+------------------+------------------+------------------+
+     * | 0x8401 | 1a28be396a14a318 |         00:00:13 |         00:07:59 |
+     * | 0x5c00 | 723ebf0d9eba3264 |         00:00:03 |         00:11:27 |
+     * | 0xe800 | ce53628a1e3f5b3c |         00:00:02 |         00:00:15 |
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5364]: Should this have a `Done` at the end like other commands do?

File: src/cli/cli.cpp [hunk 8]
```diff
@@ -5352,6 +5352,82 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+     */
+    else if (aArgs[0] == "conntime")
+    {
+        /**
+         * @cli neighbor conntime list
+         * @code
+         * neighbor conntime list
+         * 0x8401 1a28be396a14a318 age:63 conn-time:644
+         * 0x5c00 723ebf0d9eba3264 age:23 conn-time:852
+         * 0xe800 ce53628a1e3f5b3c age:23 conn-time:180
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5383]: `Done` at the end?

File: src/cli/cli.cpp [hunk 9]
```diff
@@ -5352,6 +5352,82 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+     * | 0x5c00 | 723ebf0d9eba3264 |         00:00:03 |         00:11:27 |
+     * | 0xe800 | ce53628a1e3f5b3c |         00:00:02 |         00:00:15 |
+     * @endcode
+     * @par
+     * Print the connection time and age of neighbors. Info per neighbor:
+     * - RLOC16
+     * - Extended MAC address
+     * - Last Heard (seconds since last heard from neighbor)
+     * - Connection time (seconds since link establishment with neighbor)
+     * Duration intervals are formatted as `<hh>:<mm>:<ss>` for hours, minutes, and seconds if the duration is less
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5372]: The import thinks the < and > are for HTML tags, and they end up not rendering after import.  Can you instead use curly braces

`{hh}:{mm}:{ss}`

File: src/cli/cli.cpp [hunk 10]
```diff
@@ -5352,6 +5352,82 @@ template <> otError Interpreter::Process<Cmd("neighbor")>(Arg aArgs[])
...
+     * than one day. If the duration is longer than one day, the format is `<dd>d.<hh>:<mm>:<ss>`.
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5373]: For this one, I'm assuming that one "d" should be a d.  So for this one, `{dd}d.{hh}:{mm}:{ss}`

**Review by Vyrastas — CHANGES_REQUESTED** [review]

File: src/cli/README.md [hunk 11]
```diff
@@ -2303,6 +2303,43 @@ Print table of neighbors.
 Done
 ```
 
+### neighbor conntime
+
+Print contention time and age of neighbors.
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/README.md:2308]: Ah! Thanks for noticing this. :)

**Review by Vyrastas — APPROVED** [review]

File: tests/unit/CMakeLists.txt [hunk 12]
```diff
@@ -1052,10 +1052,17 @@ add_executable(ot-test-timer
 add_executable(ot-test-toolchain
     test_toolchain.cpp test_toolchain_c.c
 )
+
 target_include_directories(ot-test-toolchain
     PRIVATE
         ${COMMON_INCLUDES}
 )
+
+target_link_libraries(ot-test-toolchain
```
Comment (jwhui) [inline tests/unit/CMakeLists.txt:1061]: Is this a related change?

Comment (abtink) [reply tests/unit/CMakeLists.txt:1061]: Yes. Now that we have `topology.cpp` file (added in this PR) we need to include the `COMMON_LIBS` when building all unit tests. We do this for other tests but I guess wanted to keep `ot-test-toolchain` simple.

File: include/openthread/instance.h [hunk 13]
```diff
@@ -53,7 +53,7 @@ extern "C" {
  * @note This number versions both OpenThread platform and user APIs.
  *
  */
-#define OPENTHREAD_API_VERSION (318)
+#define OPENTHREAD_API_VERSION (319)
```
Comment (jwhui) [inline include/openthread/instance.h:56]: Please rebase and increment again :)

**Comment by abtink** [comment]
> @abtink Is this feature currently necessary or being considered for future inclusion under network diagnostic spec for Thread Paraduxx?
> 

Yes. Devices that track this can provide this through netdiag TLVs.

> Also, can you explain [instantiating Uptime](https://github.com/openthread/openthread/blob/main/src/core/common/uptime.cpp#L51) with this large parameter [kTimer](https://github.com/openthread/openthread/blob/main/src/core/common/uptime.hpp#L127)? Do you think it should be a platform-configurable value?


This an a OT core module. It's not really related to platform.  
In short we have notion of `TimerMilli::GetTime()` which is `uin32_t` and in msec (warps every 2^32 msec or ~49 days) and provided by platform `otPlatAlramMilliGetNow()`. We want to extend it so to have a `uint64_t` value counting msec since OT instance started. That's what `Uptime` module implements.

**Comment by hemanth-silabs** [comment]
@abtink but we are starting a timer with a large value and making an inherent assumption that all platforms provide support for this large (32 bit) timeout. There might be cases where platforms support timeout which are smaller than this. It would better if the core module queries the platform about its capabilities and take that into consideration as well.

**Comment by jwhui** [comment]
> @abtink but we are starting a timer with a large value and making an inherent assumption that all platforms provide support for this large (32 bit) timeout. There might be cases where platforms support timeout which are smaller than this. It would better if the core module queries the platform about its capabilities and take that into consideration as well.

@hemanth-silabs `otPlatAlarmMilli` is a 32-bit alarm. OpenThread does not currently have any facility to support platforms with a small maximum value for `aDt`.

https://github.com/openthread/openthread/blob/4cba902db117910de22680a396b5d1c05154c12b/include/openthread/platform/alarm-milli.h#L63

**Comment by hemanth-silabs** [comment]
> > @abtink but we are starting a timer with a large value and making an inherent assumption that all platforms provide support for this large (32 bit) timeout. There might be cases where platforms support timeout which are smaller than this. It would better if the core module queries the platform about its capabilities and take that into consideration as well.
> 
> @hemanth-silabs `otPlatAlarmMilli` is a 32-bit alarm. OpenThread does not currently have any facility to support platforms with a small maximum value for `aDt`.
> 
> https://github.com/openthread/openthread/blob/4cba902db117910de22680a396b5d1c05154c12b/include/openthread/platform/alarm-milli.h#L63

@jwhui Thanks for the reply. Yes, I suppose `otPlatAlarmMilliStartAt` having a uint32 value does indicate that alarm milli works as a 32-bit alarm, but a 32 bit variable can also be used for lets say a 24 bit timer. So we were thinking, should we indicate somewhere that alarm milli **should** support UINT32_MAX inputs?

**Comment by jwhui** [comment]
> @jwhui Thanks for the reply. Yes, I suppose otPlatAlarmMilliStartAt having a uint32 value does indicate that alarm milli works as a 32-bit alarm, but a 32 bit variable can also be used for lets say a 24 bit timer. So we were thinking, should we indicate somewhere that alarm milli should support UINT32_MAX inputs?

@hemanth-silabs , yes, we can make that more explicit.

**Comment by abtink** [comment]
@hemanth-silabs, the `otPlatAlarm` APIs must support the full range of `uint32_t` which is used as parameter and return types. OT `Timer` implementation relies on the fact that it is `uint32_t`.

We can add to API docs to make this explicitly clear (avoid any confusion) but as a general rule, we expect this from any API (not just alarm APIs), i.e., it is incorrect for any platform API implementation to cast `uint32_t` param to `uint16_t` and throw away some of the bits.

If you have a platform implementation `otPlatAlaram` that wraps at 24-bit, then it is incorrect and can lead to weird and surprising bugs and behaviors. These can be hard-to-find since things may seem to work for quiet some time until issues arris when `GetNow()` wraps unexpectedly and too early.
